cyclic_boosting/quantile_matching.py

```diff
@@ -1,11 +1,161 @@
 import numpy as np
+from numpy import exp, log, sinh, arcsinh, arccosh
 from scipy.optimize import curve_fit
-from scipy.stats import norm, gamma, nbinom
+from scipy.stats import norm, gamma, nbinom, logistic
 from scipy.interpolate import InterpolatedUnivariateSpline
 
 from typing import Optional
 
 
+class J_QPD_S:
+    """
+    Implementation of the semi-bounded mode of Johnson Quantile-Parameterized
+    Distributions (J-QPD), see https://repositories.lib.utexas.edu/bitstream/handle/2152/63037/HADLOCK-DISSERTATION-2017.pdf.
+    A distribution is parameterized by a symmetric-percentile triplet (SPT).
+
+    Parameters
+    ----------
+    alpha : float
+        lower quantile of SPT (upper is ``1 - alpha``)
+    qv_low : float
+        quantile function value of ``alpha``
+    qv_median : float
+        quantile function value of quantile 0.5
+    qv_high : float
+        quantile function value of quantile ``1 - alpha``
+    l : float
+        lower bound of semi-bounded range (default is 0)
+    version: str
+        options are ``normal`` (default) or ``logistic``
+    """
+
+    def __init__(
+        self,
+        alpha: float,
+        qv_low: float,
+        qv_median: float,
+        qv_high: float,
+        l: Optional[float] = 0,
+        version: Optional[str] = "normal",
+    ):
+        if version == "normal":
+            self.phi = norm()
+        elif version == "logistic":
+            self.phi = logistic()
+        else:
+            raise Exception("Invalid version.")
+
+        self.l = l
+
+        self.c = self.phi.ppf(1 - alpha)
+
+        self.L = log(qv_low - l)
+        self.H = log(qv_high - l)
+        self.B = log(qv_median - l)
+
+        if self.L + self.H - 2 * self.B > 0:
+            self.n = 1
+            self.theta = qv_low - l
+        elif self.L + self.H - 2 * self.B < 0:
+            self.n = -1
+            self.theta = qv_high - l
+        else:
+            self.n = 0
+            self.theta = qv_median - l
+
+        self.delta = 1.0 / self.c * sinh(arccosh((self.H - self.L) / (2 * min(self.B - self.L, self.H - self.B))))
+
+        self.kappa = 1.0 / (self.delta * self.c) * min(self.H - self.B, self.B - self.L)
```

Comment:
Same comment below. In the reference given above, the kappa parameter is actually represented by alpha. I am fine with leaving kappa, but one should refer above in the function's docstring that this parameter has a different label.

Reply:
Exactly. Python doesn't like lambda ;). But you're right, let me mention it in the docstring.